test/Riok.Mapperly.Tests/Generator/IncrementalGeneratorTest.cs

```diff
@@ -0,0 +1,161 @@
+    }
+
+    [Fact]
+    public void AppendingUnrelatedTypeDoesNotRegenerateOriginal()
+    {
+        var source = TestSourceBuilder.CSharp(
+            """
+            using Riok.Mapperly.Abstractions;
+
+            namespace Test.A
+            {
+                [Mapper]
+                internal partial class FooBarMapper
+                {
+                    internal partial string FooToBar(string value);
+                }
+            }
+            """
+        );
+
+        var syntaxTree = CSharpSyntaxTree.ParseText(source, CSharpParseOptions.Default);
+        var compilation1 = TestHelper.BuildCompilation(TestHelperOptions.NoDiagnostics.NullableOption, syntaxTree);
+
+        var driver1 = TestHelper.TrackingGeneratorDriver().RunGenerators(compilation1);
+
+        var newTree = syntaxTree.WithRootAndOptions(
+            syntaxTree.GetCompilationUnitRoot().AddMembers(SyntaxFactory.ParseMemberDeclaration("struct Foo {}")!),
+            syntaxTree.Options
+        );
+
+        var compilation2 = compilation1.ReplaceSyntaxTree(compilation1.SyntaxTrees.First(), newTree);
+        var driver2 = driver1.RunGenerators(compilation2);
+
+        AssertRunResults(driver2, IncrementalStepRunReason.Cached);
+    }
+
+    [Fact]
+    public void ModifyingMapperDoesRegenerateOriginal()
+    {
+        var source = TestSourceBuilder.CSharp(
+            """
+            using Riok.Mapperly.Abstractions;
+
+            namespace Test.A
+            {
+                [Mapper]
+                internal partial class FooBarMapper
+                {
+                    internal partial string FooToBar(string value);
+                }
+            }
+            """
+        );
+
+        var syntaxTree = CSharpSyntaxTree.ParseText(source, CSharpParseOptions.Default);
+        var compilation1 = TestHelper.BuildCompilation(TestHelperOptions.NoDiagnostics.NullableOption, syntaxTree);
+
+        var driver1 = TestHelper.TrackingGeneratorDriver().RunGenerators(compilation1);
+
+        var namespaceDeclaration = syntaxTree.GetCompilationUnitRoot().Members.First() as NamespaceDeclarationSyntax;
+        var classDeclaration = namespaceDeclaration!.Members.First() as ClassDeclarationSyntax;
+        var member = SyntaxFactory.ParseMemberDeclaration("internal partial int BarToBaz(int value);")!;
+        var updatedClass = classDeclaration!.AddMembers(member);
+
+        var newRoot = syntaxTree.GetCompilationUnitRoot().ReplaceNode(classDeclaration, updatedClass);
+        var newTree = syntaxTree.WithRootAndOptions(newRoot, syntaxTree.Options);
+
+        var compilation2 = compilation1.ReplaceSyntaxTree(compilation1.SyntaxTrees.First(), newTree);
+        var driver2 = driver1.RunGenerators(compilation2);
+
+        AssertRunResults(driver2, IncrementalStepRunReason.New);
+    }
+
+    private static void AssertRunResults(GeneratorDriver driver, params IncrementalStepRunReason[] runReasons)
```

Comment:
Diagnostics should probably also be tested. FYI the `IncrementalValuesProvider` has a `WithTrackingName` method to set step tracking names.

Reply:
`WithTrackingName` is not available in older versions of roslyn. I could use the auto generated name or wrap everything in `#if` preproccessors

Reply:
No idea how to use `WithTrackingName` with older roslyn versions

Reply:
You could probably just add an extension method `IncrementalValueProvider<TSource> WithTrackingName<TSource>(this IncrementalValueProvider<TSource> source, string name) => source;` to mock this for older roslyn versions.

Reply:
My main issue is getting the tests to pass. I can use preprocessors/mock `WithTrackingName`, but I can't get the tests to pass.

For instance if I wrap the tests in Net_7_And_Above the tests will still run for a version of Roslyn that does not add the tracking name

Reply:
Hmmm, but the unit tests only run for the latest Roslyn version supported by Mapperly, don't they?

Reply:
That's what I thought too 😆 

I'll double check I probably made a nistake